### ISSUE TITLE
High-value kitty stealing: minOfferPrice is not communicated on-chain by user

### ISSUE THREAD

**Issue by TomLeeFounder** [dapperlabs/cryptokitties-bounty-2#22]
### Description

This is dangerous for high value kitties.

According the mechanics described in the ReadMe, the only on-chain transaction done by the user accepting an offer is for them to call `KittyCore.approve()`, passing in `tokenID` and the address of Offers.sol as arguments.

However, this means that minOfferPrice is communicated to CK off-chain.

If minOfferPrice is altered while in transit off-chain, CK's worker may call `.fulfillOffer()` with a much lower minOfferPrice than the user intended.

### Scenario

If an attacker called `.cancelOffer()` in between when the user calls `.approve()` and when CK's worker calls `.fulfillOffer()`, and minOfferPrice was altered in transit, the attacker could ensure that a high-value kitty is sold for a low price.

### Impact

Medium Severity: High Impact if it occurs, Low Probability of occurring. This is especially lucrative for high value kitties (some kitties sell for $100,000). 

### Reproduction

(1) Attacker sends low offer of 0.01 ETH on Kitty\#1.
(2) Attacker sends a higher offer of 100 ETH on Kitty\#1 from a different account.
(3) Owner accepts 100 ETH offer through CK UX. This calls `.approve()` on Kitty\#1 for `Offers.sol`, and tries to send minOfferPrice off-chain to CK
(4) Attacker alters off-chain communication of minOfferPrice, changing it from 100 ETH to 0.01 ETH.
(5) Attacker calls `.cancelOffer()` on their 100 ETH offer
(6) CK's worker detects confirmation of Owner's `.approve()` call, sees the off-chain communication of 0.01 ETH, and sends `.fulfillOffer()` for Kitty\#1 with minOfferPrice of 0.01.
(7) Attacker receives Kitty\#1 and pays Owner only 0.01 ETH.

### Fix

Four potential fixes:

Fix\#1 (most cumbersome):
Remove the ability for COO to call `.fulfillOffer()` on behalf of users so that they can be sure that they can communicate their minOfferPrice.

Fix\#2 (somewhat cumbersome):
In `Offers.sol`, add an on-chain function `.revokeCOOpermissionForToken()` that allows users to specify that COO cannot fulfill offers for a particular token number.

Fix\#3 (less cumbersome)
Give users the option in the UX to *not* have COO try to fulfill their offer, but require them to sign this off-chain message with their key (as is done for first signing into CK). This ensures that the Attacker cannot change the message in transit, and it lets them be sure that they are the ones who call `.fullfillOffer()`.

Fix\#4 (least cumbersome):
Give users the option in the UX to *not* have COO try to fulfill their offer, but don't require them to sign it with their key. (This doesn't prevent much, since Attacker can simply alter this off-chain communication as well).

**Comment by sunsetlover** [comment]
I'm honestly confused why they even let the COO fulfill offers. Seems like a bad choice to make.

**Comment by sunsetlover** [comment]
Yeah I'm curious as to why they decided to allow the COO to fulfill offers. I don't understand the rational. So that people don't have to submit their own txs?

**Comment by hwrdtm** [comment]
Thanks @TomLeeFounder for your feedback! We have acknowledged it and will be taking it into consideration.

**Comment by hashkitty** [comment]
Do I understand correctly that attack is not working exactly as described in Reproduction? If attacker cancels offer in step 5 and does not submit new offer before step 6 then fulfillOffer should fail? This would also require a race condition between approve transaction and order recreation to look more realistic. If approve (step 3) is mined before offer is recreated (step 5) then fraud is clearly detectable from the logs - approve comes before cancel. If step 5 transaction is mined before step 3 approve transaction, then fraud indication is not that clear since it may happen in legitimate scenario.

I assume that offer recreation with smaller price in short period of time should always fail  CK's "off-chain checks" if COO behaves as intended.

**Comment by arthcmr** [comment]
Thank you for the report @TomLeeFounder. We don't consider this a bug per se because it was part of the original design. However, we acknowledge that there are centralization questions with this choice of architecture, and we think your suggestions are quite valuable. So we have decided to still reward you with 50 points for this Note.
